tools/lkl/lklfuse.c

```diff
@@ -545,12 +547,119 @@ const struct fuse_operations lklfuse_ops = {
 	.fallocate = lklfuse_fallocate,
 };
 
+static int lklfuse_parse_vfs_flags(bool ro, const char *opts, int *flags,
+				   char *remain, size_t remlen)
+{
+	size_t optslen;
+	char tokbuf[4096];
+	char *tb = tokbuf;
+	char *tok;
+	char *saveptr = NULL;
+	/* we should consider using libmount for this mapping */
+	struct flagmap {
+		const char *opt;
+		int flag;
+		bool invert;
+	} map[] = {
+		{ "ro", LKL_MS_RDONLY },
+		{ "nosuid", LKL_MS_NOSUID },
+		{ "nodev", LKL_MS_NODEV },
+		{ "noexec", LKL_MS_NOEXEC },
+		{ "sync", LKL_MS_SYNCHRONOUS },
+		{ "remount", LKL_MS_REMOUNT },
+		{ "dirsync", LKL_MS_DIRSYNC },
+		{ "nosymfollow", LKL_MS_NOSYMFOLLOW },
+		{ "noatime", LKL_MS_NOATIME },
+		{ "nodiratime", LKL_MS_NODIRATIME },
+		{ "bind", LKL_MS_BIND },
+		{ "rbind", LKL_MS_BIND | LKL_MS_REC },
+		{ "move", LKL_MS_MOVE },
+		{ "silent", LKL_MS_SILENT },
+
+		{ "relatime", LKL_MS_RELATIME },
+		{ "iversion", LKL_MS_I_VERSION },
+		{ "strictatime", LKL_MS_STRICTATIME },
+		{ "lazytime", LKL_MS_LAZYTIME },
+
+		/* opts below aren't passed to external mount.type helpers */
+		{ "unbindable", LKL_MS_UNBINDABLE },
+		{ "runbindable", LKL_MS_UNBINDABLE | LKL_MS_REC },
+		{ "private", LKL_MS_PRIVATE },
+		{ "rprivate", LKL_MS_PRIVATE | LKL_MS_REC },
+		{ "slave", LKL_MS_SLAVE },
+		{ "rslave", LKL_MS_SLAVE | LKL_MS_REC },
+		{ "shared", LKL_MS_SHARED },
+		{ "rshared", LKL_MS_SHARED | LKL_MS_REC },
+
+		/* opts below invert flags */
+		{ "rw", LKL_MS_RDONLY, true },
+		{ "suid", LKL_MS_NOSUID, true },
+		{ "dev", LKL_MS_NODEV, true },
+		{ "exec", LKL_MS_NOEXEC, true },
+		{ "async", LKL_MS_SYNCHRONOUS, true },
+		{ "symfollow", LKL_MS_NOSYMFOLLOW, true },
+		{ "atime", LKL_MS_NOATIME, true },
+		{ "diratime", LKL_MS_NODIRATIME, true },
+		{ "loud", LKL_MS_SILENT, true },
+		{ "norelatime", LKL_MS_RELATIME, true },
+		{ "noiversion", LKL_MS_I_VERSION, true },
+		{ "nostrictatime", LKL_MS_STRICTATIME, true },
+		{ "nolazytime", LKL_MS_LAZYTIME, true },
+	};
+	*flags = ro ? LKL_MS_RDONLY : 0;
+
+	optslen = 0;
+	if (opts)
+		optslen = strlen(opts);
+	if (optslen == 0)
+		return 0;
+	/* need enough space in @remain to at most pass through all options */
+	if (optslen >= remlen || optslen >= sizeof(tokbuf))
+		return -EINVAL;
+
+	*remain = '\0';
+	strncpy(tokbuf, opts, sizeof(tokbuf));
+	while ((tok = strtok_r(tb, "\\,", &saveptr)) != NULL) {
```

Comment:
Do we need \ as a separator here? Do you have some usage examples?

Reply:
Thanks for taking a look at the changes...

> Do we need \ as a separator here?

LKL kernel mount options provided via the `lklfuse -opts=...` parameter are separated with '\,' instead of ',' so that they don't get processed as part of the regular `fuse_opt_parse()` process.
Using `strtok(delim="\\,")` isn't ideal in that we should be handling each two-char '\,' (and '\,') sequence, instead of accepting either as a delimiter, but it does simplify parsing. I'll have a play with using `strstr()` to do things properly.

> Do you have some usage examples?

```
lklfuse -f -o 'fsname=/dev/zram0,subtype=lkl.xfs,type=xfs,opts=rw\,noatime\,filestreams' /dev/zram0 /mnt/test
```

In this case the escaped `opts` parameter value goes through to `lklfuse_parse_vfs_flags()`, where `rw` and `noatime` are consumed as VFS flags (`rw` would invert any prior `ro`); `filestreams` is passed through as a string to `lkl_sys_mount()` and handled as an xfs specific option in `xfs_super.c`.

Reply:
Got it now. If you can add a comment block explaining this I am ok with merging this as it is.

Reply:
hmm, your suspicions were correct: although the `opts` escapes are needed, they helpfully get weeded out by libfuse (`process_real_option_group()`) before being processed here, so we don't need to look for `\`.